radix-engine/src/engine/track.rs

```diff
@@ -456,31 +461,51 @@ impl<'s, S: SubstateStore> Track<'s, S> {
         }
     }
 
-    pub fn resource_manager_invoke(
+    pub fn borrow_global_mut_resource_manager(
         &mut self,
-        resource_address: &ResourceAddress,
-        function: &str,
-        args: Vec<ScryptoValue>,
-    ) -> Result<Option<Bucket>, RuntimeError> {
-        if let Some(substate_update) = self.resource_managers.remove(resource_address) {
-            let mut resource_manager: ResourceManager = substate_update.value;
-            let result = resource_manager
-                .main(resource_address.clone(), function, args, self)
-                .map_err(RuntimeError::ResourceManagerError)?;
+        resource_address: ResourceAddress,
+    ) -> Result<ResourceManager, RuntimeError> {
+        let maybe_resource = self.resource_managers.remove(&resource_address);
+        if let Some(SubstateUpdate { value, prev_id }) = maybe_resource {
+            self.borrowed_resource_managers
+                .insert(resource_address, prev_id);
+            Ok(value)
+        } else if self
+            .borrowed_resource_managers
+            .contains_key(&resource_address)
+        {
+            panic!("Invalid resource manager reentrancy");
```

Comment:
Should this be an error instead of panic? Would also consider moving it up for readability.

Reply:
No, panic for now since this logic is all enforced by system at the moment